### PR TITLE
sw: Multi-cluster runtime improvements

### PR DESCRIPTION
**Pull request by fischeti**
This PR aims to improve the snitch runtime when when the snitch cluster is used as a dependency in a multi-cluster system.

### Changes

* Autogenerate `snitch_cluster_defs.h` with the number of clusters
* Getting rid of `nr_quadrants` in configuration, which is obsolete in my opinion and was never actually used anywhere.

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -35,7 +35,9 @@ snitch-cluster-sw:
   artifacts:
     paths:
       - sw/math/include/bits/alltypes.h
+      - target/snitch_cluster/sw/runtime/common/*.h
       - target/snitch_cluster/sw/**/build/*.elf
+      - target/snitch_cluster/sw/runtime/rtl/build/libsnRuntime.a
```
Comment (colluca) [inline .gitlab-ci.yml:40]: Why is this needed? Is there any dependent CI stage job which uses these?

File: docs/schema/snitch_cluster.schema.json [hunk 2]
```diff
@@ -14,6 +14,11 @@
             "description": "Optional name for the generated wrapper.",
             "default": "snitch_cluster"
         },
+        "nr_clusters": {
+            "type": "number",
+            "description": "Number of clusters. Only relevant for the SW runtime.",
+            "default": 1
+        },
```
Comment (colluca) [inline docs/schema/snitch_cluster.schema.json:21]: Why is this added as a property of the Snitch cluster?

**Review by colluca — CHANGES_REQUESTED** [review]
Since `snitch_cluster_defs.h` is also auto-generated now, perhaps we want to just merge it with `snitch_cluster_cfg.h`. I believe the previous idea was to just separate the auto-generated defines to a single place, and everything else derived from it. But I think it's not too relevant and we can just merge them to possibly simplify the codebase.

Further minor comments follow.

File: .gitlab-ci.yml (hunk 1 above)
Comment (fischeti) [reply .gitlab-ci.yml]: There was an issue, but I don't remember anymore why it failed. I will check again

File: target/common/test/tb_lib.hh [hunk 3]
```diff
@@ -126,7 +126,6 @@ struct BootData {
     uint64_t global_mem_start;
     uint64_t global_mem_end;
     uint32_t cluster_count;
-    uint32_t s1_quadrant_count;
```
[the remark this answers is not in the document]
Comment (fischeti) [reply target/common/test/tb_lib.hh:129]: No, I haven't checked that. But doesn't Occamy use a different configuration file resp. schema anyway?

[the remark this answers is not in the document]
Comment (colluca) [reply target/common/test/tb_lib.hh:129]: Yeah I think we can remove this Occamy-specific parameter from this repo, and later align Occamy if needed.

File: .gitlab-ci.yml (hunk 1 above)
Comment (fischeti) [reply .gitlab-ci.yml]: I remember the issue again. The [`run.py`](https://github.com/pulp-platform/snitch_cluster/blob/17aa42a7a002c633e91b5948a94b71404d1bd5d9/target/snitch_cluster/sw/apps/blas/gemm/test/run.py#L40) script builds the software including runtime again, but it does not generate the header files again (see [here](https://iis-git.ee.ethz.ch/github-mirror/snitch_cluster/-/jobs/1350489))

Comment (colluca) [reply .gitlab-ci.yml]: I see, thanks. I will look into this further, and otherwise we can merge it as is.

Comment (fischeti) [reply .gitlab-ci.yml]: I reverted the commit to check it, but let me know if I should rerevert it to merge it😉

Comment (colluca) [reply .gitlab-ci.yml]: Thanks Tim :) I think we will be able to fix this in https://github.com/pulp-platform/snitch_cluster/pull/137 where we are anyways slightly generalizing the `run.py` script. So no need to revert it now, I think we can get this PR to pass as is once we rebase on the other one.
